pkg/virt-launcher/virtwrap/network/podnic_test.go

```diff
@@ -0,0 +1,155 @@
+		vmi                           *kubevirtv1.VirtualMachineInstance
+		newPodNetworkConfiguratorMock = func(*podNIC) (infraconfigurators.PodNetworkInfraConfigurator, error) {
+			return mockPodNetworkConfigurator, nil
+		}
+		newPodNICWithMocks = func(vmi *v1.VirtualMachineInstance) *podNIC {
+			podnic, err := newPodNIC(vmi, &vmi.Spec.Networks[0], mockNetwork, cacheFactory, nil)
+			Expect(err).ToNot(HaveOccurred())
+			podnic.podNetworkConfiguratorFactory = newPodNetworkConfiguratorMock
+			podnic.podInterfaceName = primaryPodInterfaceName
+			return podnic
+		}
+	)
+	BeforeEach(func() {
+		cacheFactory = fake.NewFakeInMemoryNetworkCacheFactory()
+
+		ctrl = gomock.NewController(GinkgoT())
+		mockNetwork = netdriver.NewMockNetworkHandler(ctrl)
+		mockPodNetworkConfigurator = infraconfigurators.NewMockPodNetworkInfraConfigurator(ctrl)
+	})
+	AfterEach(func() {
+		ctrl.Finish()
+	})
+	Context("when pod networking setup fails", func() {
+		BeforeEach(func() {
+			address1 := &net.IPNet{IP: net.IPv4(1, 2, 3, 4)}
+			address2 := &net.IPNet{IP: net.IPv4(169, 254, 0, 0)}
+			fakeAddr1 := netlink.Addr{IPNet: address1}
+			fakeAddr2 := netlink.Addr{IPNet: address2}
+			mockNetwork.EXPECT().ReadIPAddressesFromLink(primaryPodInterfaceName).Return(fakeAddr1.IP.String(), fakeAddr2.IP.String(), nil)
+			mockNetwork.EXPECT().IsIpv4Primary().Return(true, nil).Times(1)
+			mockPodNetworkConfigurator.EXPECT().DiscoverPodNetworkInterface(primaryPodInterfaceName).Times(1)
+			mockPodNetworkConfigurator.EXPECT().GenerateDHCPConfig().Return(&cache.DHCPConfig{}).Times(1)
+			mockPodNetworkConfigurator.EXPECT().GenerateDomainIfaceSpec().Times(1)
+			mockPodNetworkConfigurator.EXPECT().PreparePodNetworkInterface().Return(fmt.Errorf("podnic_test: forcing prepare networking failure")).Times(1)
+		})
+		It("should return a CriticalNetworkError at phase1", func() {
+
+			domain := NewDomainWithBridgeInterface()
+			vmi := newVMIBridgeInterface("testnamespace", "testVmName")
+
+			api.NewDefaulter(runtime.GOARCH).SetObjectDefaults_Domain(domain)
+
+			podnic := newPodNICWithMocks(vmi)
+			//podnic.launcherPID = &pid
+			err := podnic.PlugPhase1()
+			Expect(err).To(HaveOccurred(), "SetupPhase1 should return an error")
+
+			_, ok := err.(*neterrors.CriticalNetworkError)
+			Expect(ok).To(BeTrue(), "SetupPhase1 should return an error of type CriticalNetworkError")
+		})
+	})
+	Context("when DHCP startup fails", func() {
+		BeforeEach(func() {
+			mockNetwork.EXPECT().StartDHCP(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).Return(errors.New("podnic_test: forcing failure at DHCP start"))
+		})
+		It("phase2 should panic", func() {
+			testDhcpPanic := func() {
+				launcherPID := 1
+				domain := NewDomainWithBridgeInterface()
+				vmi := newVMIBridgeInterface("testnamespace", "testVmName")
+				api.NewDefaulter(runtime.GOARCH).SetObjectDefaults_Domain(domain)
+				podnic := newPodNICWithMocks(vmi)
+				podnic.launcherPID = &launcherPID
```

Comment:
do we need launcher while testing phase2 ?

Reply:
Nope, removed.